File: src/modelo/EstoqueServico.java
package modelo;

import pojo.*;
import java.util.ArrayList;
import java.util.List;

/**
 * SERVIÇO 2 — EstoqueServico
 * Classe de modelo que gerencia estoques e movimentações de produtos.
 */
public class EstoqueServico {

    private List<Estoque> estoques = new ArrayList<>();
    private int proximoId = 1;

    // ── CRUD de estoques ─────────────────────────────────────────────────────

    public Estoque criarEstoque(String local) {
        Estoque e = new Estoque(proximoId++, local);
        estoques.add(e);
        System.out.println("[ESTOQUE] Criado: " + e);
        return e;
    }

    public Estoque buscarEstoquePorId(int id) {
        return estoques.stream()
                       .filter(e -> e.getId() == id)
                       .findFirst()
                       .orElse(null);
    }

    // ── Movimentações ────────────────────────────────────────────────────────

    /**
     * Adiciona um produto a um estoque específico.
     */
    public boolean entradaProduto(int idEstoque, Produto produto) {
        Estoque estoque = buscarEstoquePorId(idEstoque);
        if (estoque == null) {
            System.out.println("[ERRO] Estoque " + idEstoque + " não encontrado.");
            return false;
        }
        estoque.adicionarProduto(produto);
        System.out.println("[ENTRADA] " + produto.getNome() +
                           " → Estoque: " + estoque.getLocal());
        return true;
    }

    /**
     * Remove um produto de um estoque.
     */
    public boolean saidaProduto(int idEstoque, int idProduto) {
        Estoque estoque = buscarEstoquePorId(idEstoque);
        if (estoque == null) return false;
        boolean ok = estoque.removerProduto(idProduto);
        if (ok) System.out.println("[SAÍDA] Produto id=" + idProduto +
                                   " removido do estoque " + estoque.getLocal());
        return ok;
    }

    // ── Relatórios ───────────────────────────────────────────────────────────

    /**
     * Verifica se alguma vacina perecível está vencida em qualquer estoque.
     */
    public void alertarVencidos() {
        System.out.println("\n=== ALERTA DE PRODUTOS VENCIDOS ===");
        boolean achou = false;
        for (Estoque est : estoques) {
            for (Produto p : est.getProdutos()) {
                if (p instanceof Perecivel) {
                    Perecivel per = (Perecivel) p;
                    if (per.isVencido()) {
                        System.out.println("  [VENCIDO] " + p.getNome() +
                                           " | validade: " + per.getDataValidade() +
                                           " | estoque: " + est.getLocal());
                        achou = true;
                    }
                }
            }
        }
        if (!achou) System.out.println("  Nenhum produto vencido encontrado.");
        System.out.println("===================================\n");
    }

    /**
     * Imprime relatório completo de todos os estoques.
     */
    public void relatorioGeral() {
        System.out.println("\n========= RELATÓRIO GERAL DE ESTOQUES =========");
        for (Estoque est : estoques) {
            System.out.println("\nEstoque: " + est.getLocal() +
                               " (id=" + est.getId() + ") — " +
                               est.getTotalProdutos() + " produto(s)");
            for (Produto p : est.getProdutos()) {
                System.out.println("  > " + p);
            }
        }
        System.out.println("================================================\n");
    }

    public List<Estoque> listarEstoques() { return new ArrayList<>(estoques); }
}

File: servidor-cpp/src/servico/EstoqueServico.cpp
#include "EstoqueServico.hpp"


EstoqueServico::EstoqueServico() {}

Estoque EstoqueServico::criarEstoque(const string& local)
{
    Estoque e(proximoId++, local);
    estoques.push_back(e);
    cout << "[ESTOQUE] Criado: " << e.toString() << endl;
    return e;
}

Estoque* EstoqueServico::buscarEstoquePorId(int id)
{
    // para cada estoque na lista de estoques, verifica se o id corresponde ao id buscado
    for (auto& estoque : estoques)
    {
        if (estoque.getId() == id)
        {
            return &estoque;
        }
    }
    return nullptr;
}

bool EstoqueServico::entradaProduto(int estoqueId, Produto* p)
{
    Estoque* estoque = buscarEstoquePorId(estoqueId);
    if (estoque)
    {
        estoque->adicionarProduto(p);
        cout << "[ENTRADA] " << p->getNome() << 
        " -> Estoque: " << estoque->getLocal() << endl;
        return true;
    }
    return false;
}


bool EstoqueServico::saidaProduto(int estoqueId, int ProdutoId)
{
    Estoque* estoque = buscarEstoquePorId(estoqueId);

    if (!estoque)
    {
        return false;
    }
    

    bool removido = estoque->removerProduto(ProdutoId);
    if (removido)
    {
        cout << "[SAIDA] Produto id=" << ProdutoId << " removido do Estoque " << estoque->getLocal() << endl;
    }
    return removido;
}

void EstoqueServico::alertarVencidos()
{
    cout << "\n========= ALERTA DE PRODUTOS VENCIDOS =========" << endl;
    bool encontrou = false;

    for (const auto& estoque : estoques)
    {
        for(auto* produto : estoque.getProdutos()){
            const Perecivel* p = dynamic_cast<const Perecivel*>(produto);
            if (p && p->ehVencido())
            {
                cout << " [VENCIDO] " << produto->getNome() << 
                " | validade: " << p->getDataValidade() 
                << " | Estoque: " << estoque.getLocal() << endl;
                encontrou = true;
            }
        }
    }

    if (!encontrou)
    {
        cout << "[ALERTA] Nenhum produto vencido encontrado." << endl;
    }
}

void EstoqueServico::relatorioGeral()
{
    cout << "\n========= RELATORIO GERAL DE ESTOQUES =========" << endl;
    for (const auto& estoque : estoques)
    {
        cout << "\nEstoque: " << estoque.getLocal()
        << " (id=" << estoque.getId() << ") - "
        << estoque.getTotalProdutos() << " produto(s)"
        << endl;

        for (const auto& produto : estoque.getProdutos())
        {
            cout << "  > " << produto->toString() << endl;
        }
    }
    cout << "===============================================" << endl;
}

const vector<Estoque>& EstoqueServico::listarEstoques() const
{
    return estoques;
}